### PR TITLE
Clicking scoreboard shows better defaults in options

### PR DESCRIPTION
**Pull request by 96LawDawg**
When you click a scoreboard currently, the overlay values default to the player in the first seat and the first round.  This PR changes that so that the default value is the playerName.

I want to change the default round to be the next round.  But I wanted to go ahead and get this started because I had so much trouble doing it.  Hopefully the round will be easier.

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1700/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1700-pr-test`.

**Comment by rogerl50** [comment]
> But that is worse than the default. If you have say 3 people entering scores, this works for the first person, but then because there is an entry in the next round, for the second person it increases another round and so on. What needs to happen is if there are no entries for round x in any of the seats, then it works like I tried (the default is the next/upcoming round). But if there are any values for the current round in a seat, then the value should be rounds.length-1. But I don't know how to do the check. Anybody want to help out?

I'm not sure that this is possible without lots of code. It is very similar to the problem of preventing multiple people from pressing a button.

**Comment by ArnoldSmith86** [comment]
> Anybody want to help out?

Can you show examples of how the score properties in each seat look like for each scenario?

**Comment by 96LawDawg** [comment]
Scenario: 3 players just finished round 4.  First player, LawDawg clicks on the scoreboard and gets this:

Image1

If the other players click the scoreboard right away, they will see the same thing (except with the change that is working in this PR, they will see their own name).  But if LawDawg is fast to enter his score then the next player to click on the scoreboard sees this:

Image2

The round has advanced to the 5 as the default, but if this were working as I think it should, the round would stay 4 until all 3 players have entered a score for that round.  Like this:

Image3

So for the first player, ` value: rounds.length` works.  For the other players, it should be ` value: rounds.length-1` as the default.  The problem for me is devising the check for when you should use one or the other.

**Comment by 96LawDawg** [comment]
And you already fixed it before I answered.  The rounds now work.  Thank you.

So one thing my original commit changes from production is that if you are not in a seat, you can't click on the scoreboard.  That wasn't intentional, but I don't have a problem with that.  It won't change the way games are played because you have to be in a seat to make scoreboards work anyway.  I will need to tweak a few words on the tutorial.  But first I wanted to see if anybody else has a problem with that.

**Comment by robartsd** [comment]
> So for the first player, ` value: rounds.length` works. For the other players, it should be ` value: rounds.length-1` as the default. The problem for me is devising the check for when you should use one or the other.

Should it just be `rounds.length - 1` unless there is already a score defined for that position for the current player?

**Comment by ArnoldSmith86** [comment]
> So one thing my original commit changes from production is that if you are not in a seat, you can't click on the scoreboard.

But... why? You only changed defaults of inputs.

**Comment by 96LawDawg** [comment]
YOU are asking ME?  I have no idea.  You built the click functionality.

On production, load the tutorial for Scoreboard, basic variant.  Click on the top-right scoreboard (the wide one).  You aren't in a seat but you get the overlay.  Now do it in this PR.  The overlay doesn't open.

**Comment by ArnoldSmith86** [comment]
I have an idea. Will check tomorrow.

File: client/js/widgets/scoreboard.js
Comment (robartsd) [inline client/js/widgets/scoreboard.js:66]: Could have an issue here if a seat has something other than an array in the scoreProperty.

**Comment by robartsd** [comment]
> YOU are asking ME? I have no idea. You built the click functionality.
> 
> On production, load the tutorial for Scoreboard, basic variant. Click on the top-right scoreboard (the wide one). You aren't in a seat but you get the overlay. Now do it in this PR. The overlay doesn't open.

Is the issue that you are not in a seat or is the issue that none of the seats are occupied?

File: client/js/widgets/scoreboard.js
Comment (robartsd) [reply client/js/widgets/scoreboard.js:66]: Also, reduce function will fail if seats is an empty array (can fix by providing an initial value to reduce).

**Comment by 96LawDawg** [comment]
> Is the issue that you are not in a seat or is the issue that none of the seats are occupied?

All of the seats but one in the tutorial room are occupied.  So I think the issue is that the player clicking is not in a seat.

File: client/js/widgets/scoreboard.js
Comment (robartsd) [inline client/js/widgets/scoreboard.js:83]: If `playerName` is not found, then `players.find()` returns `undefined`. Trying to access the `value` property of `undefined` is an error.

File: client/js/widgets/scoreboard.js
Comment (96LawDawg) [reply client/js/widgets/scoreboard.js:66]: > Could have an issue here if a seat has something other than an array in the scoreProperty.

I don't think so. If it isn't an array, then there is just a totals line and no rounds.  And the change still works when there is just a totals line.

**Comment by 96LawDawg** [comment]
The latest commit by @robartsd fixes the problem.  But I just found another problem.  We need a solution that works with team scoring.

File: client/js/widgets/scoreboard.js
Comment (96LawDawg) [reply client/js/widgets/scoreboard.js]: I tried something like
`players.find(player => player.text == playerName)?.value ?? null`.  I changed null to ' ', players[0].value, and several other things trying to fix the team problem, but none of that worked.

Comment (robartsd) [reply client/js/widgets/scoreboard.js]: It looks like `players.find(player => player.text == playerName)?.value` should return `undefined` when `playersName` is not found (equivalent of what my code does. Not sure what `showInputOverlay` does with `null` or `''`, but `undefined` should behave as if `value` was not set at all.

**Comment by ArnoldSmith86** [comment]
> We need a solution that works with team scoring.

Was this addressed by the latest commit?

**Comment by 96LawDawg** [comment]
> > We need a solution that works with team scoring.
> 
> Was this addressed by the latest commit?

No.  Current status is this does not work with teams or when all seats are empty.

**Comment by 96LawDawg** [comment]
Just to be clear, there are 2 different errors.  If the seats are empty, error is "Reduce of empty array with no initial value." If using teams, the error is "seats.map is not a function."  Right now, the log errors line is not even triggering before the client freezes.

**Comment by robartsd** [comment]
> Just to be clear, there are 2 different errors. If the seats are empty, error is "Reduce of empty array with no initial value." If using teams, the error is "seats.map is not a function." Right now, the log errors line is not even triggering before the client freezes.

Thanks for the clarification. My last commit addressed the default selection when using teams, but not problems with everyPlayerFilledLastRound. I'm thinking we should add a class method to check if a particular round has values for all seats and call that method with the last round.

**Review by rogerl50 — APPROVED** [review]

**Comment by 96LawDawg** [comment]
Going to approve this in a couple of days pending any objections.

**Pull request merged**
